### DISCUSSION TITLE
TEMP again

### DISCUSSION THREAD

**Discussion by bnilsson11** [ra3xdh/qucs_s#994] [Q&A] [answered]
I have a medium complex circuit, with R, C, Diodes and transistors. No IC's for the moment.
I want to test the temperature dependence, so I add an .OPTIONS and set a TEMP value in it.
If I "calculate DC bias" I get a reasonable value for EXACTLY 27 degrees ONLY.
E.g. 26.85 or any other value than 27 gives very strange behaviour.
Can this be explained?
I do not have a clue how to debug this, help will be appreciated.
[MFA1-debug.sch.zip](https://github.com/user-attachments/files/17369290/MFA1-debug.sch.zip)

**Comment by tomhajjar** [comment]
Try using the two ways to do Temp Sweep. Original ngspice method used "TEMPER". New method uses UseGlobTemp=yes.

MFA1 out offset voltage over Temp doesn't look correct. Transient does...

![2024-10-14_195751](https://github.com/user-attachments/assets/ecf857d2-d3f9-4c44-9061-ed065967a4f9)
![2024-10-14_195733](https://github.com/user-attachments/assets/afc56d91-5f06-4893-b305-ef10203c55df)
![2024-10-14_201503](https://github.com/user-attachments/assets/8a5d61bd-d9aa-47fd-9dec-c54a86a35a2d)

[Temperature_Sweep_prj.zip](https://github.com/user-attachments/files/17370429/Temperature_Sweep_prj.zip)
[MFA1.zip](https://github.com/user-attachments/files/17370553/MFA1.zip)

**Comment by bnilsson11** [comment]
The result you get for the temperature sweep for MFA1-debug.sch seems not correct.
Please do a "DC bias" without any TEMP set or sweep.
You should find that out≈46.9mV.  I_PNP and Ipr3 ≈ 0.46A in this case.
As soon as you add .OPTIONS and set TEMP to ≠ 27 (e.g. 27.1)  or sweep it, Out≈1.5V and I_PNP ≈ 3.56A. Not normal.
I see this problem only in this particular circuit.

However, in my log it always says (here I use TEMP = 27.1):

> Doing analysis at TEMP = 27.100000 and TNOM = 27.000000

I see this in all my simulations.
This value of TNOM is unexpected to me. To my taste, it should be 26.85.
I have no knowledge where "TNOM = 27" is defined.

If it is different at your site, it is interesting.

**Comment by tomhajjar** [comment]
I did do the "DC bias. You can press F8 in Windows.

1. What is the source of the schematic? It's an odd design using two separate feedback paths for the full complementary differential inputs. Normally a single feedback network is used like in the Leach amp. This type of input dates back to the 1970s for the Southwest Technical Products amps as shown below.
https://leachlegacy.ece.gatech.edu/lowtim/graphics/ckt.pdf

2. Where did you get the devices? The parameters are different than the built in BJT Extended Library.

![RE_Mar_1973_pg42](https://github.com/user-attachments/assets/cdcd1989-22fb-4594-a758-dc5b522763bf)

**Comment by bnilsson11** [comment]
The design was made by Tim de Paravacini for Musical Fidelity in the early eighties, released as "A1 Integrated Class A Amplifier".
![musical_fidelity_a1_stereo_integrated_amplifier](https://github.com/user-attachments/assets/59aa5c74-099f-4f19-9953-978b40da0b30)
It has been described in detail at https://www.markhennessy.co.uk/mf_a1/

It has achieved a somewhat "cult" status, and was actually re-released this year, see https://www.musicalfidelity.com/products/a1/a1-2

Clones with a credit-card sized PCB's can be bought at eBay.

The original had what I would sa a "questionable" thermal design, the to surface approaching 70C, almost dangerous, I don't know if this is the case with the new product. Anyway, if you believe that amplifiers may sound different, this is an example of a very good sounding amp. The design has a few drawbacks, that may or may not be important when you play music and not test signals.
It seems to be an accepted consensus not to have capacitors in the feedback loop, and this design certainly has it. The idling current is controlled by applying the voltage over the sense resistors as 1:1 DC feedback to the negative diff pair input. For AC, this feedback is decoupled by a resistor and cap to a level of -25dB, equal on both positive and negative sides.
The main objection to this is that when you play high level signals over some time, 1-2 seconds, the load current to the speaker passes thru the sense resistors and adds to the DC feedback voltage. Long time high level then changes/suppresses the setpoint for the idling current  and eventually you get severe crossover distorsion. The time constant for this is about 2 seconds. When you play music, however, this is rarely an issue. Unless you play dynamically compressed death metal at levels that would blow your roof off.

My project is to investigate if this design can be improved in some way. Feedback without caps, linearising the input diff pairs, adding current mirrors and constant current sources, etc.  Since this is a class Ab (note uppercase A, lowercase B) the thermal behaviour is important. 
This is why it is so frustrating that the .OPTIONS TEMP parameter is not working for me.
The fact that TEMP=27 is working and TMP=27.1 is not, tells me that there is something deeper in the simulation than just an unexpected temperature dependence in the models.

My devices came from various sources. 
I have to admit I did not use BJT_Extended, mainly because the device list is unsorted, not so easy to find.
I did a text search in the BJT_Extended.lib and found them all, thanks for the suggestion.
However, it did not change anything. TEMP 27 works (0.5A idling current, 16mV DC out), while TEMP 27.1 or 26.85 did not (4A idling current, 5.6V DC out).

**Comment by bnilsson11** [comment]
I made a new schematic from scratch, same content as the one posted.
Now it worked, no TEMP≠27 kinks.
The original schematic must have been corrupt in some unknown way.
Case could be closed.

But I still wonder where TNOM=27 is set....

**Comment by tomhajjar** [comment]
Can you post the schematic? You could direct your question to @ra3xdh 

From my experience with solid state audio power amplifiers, thermal stability and sufficiently large heatsinks are critical if you want the amplifier to survive long sessions at high volumes. The MFA1 doesn't have any temperature compensation circuitry and I cannot vouch for the heat sink size. Amplifier stability is also critical if you want the tweeters in your speakers to survive. 

I have blown up too many amplifiers and tweeters to ever want to own a badly designed amplifier again. Amps that worked fine for normal listening would self destruct during a long party. The last amp I built was a clone of the Quad 405 current dumping amplifier. There are dozens of variations of current dumping amps on diyAudio.

![Quad 405-Douglas_Self](https://github.com/user-attachments/assets/bd73bd14-74cd-4e27-849c-d93f3ebe519a)

**Comment by bnilsson11** [comment]
I might put myself on thin ice here, but I think the temperature compensation is not as critical if you use regular current measurement feedback to control the idling current, provided that the heatsinking is sufficient.  The A1 is rather unique in this respect.
I post the original design, with 2N3055/2955 as output transistors.
Also AC gain, and two TRAN showing the output to 4ohm load at 3ms and also after 5 seconds high power.
It is obvious that the measured load current has suppressed the bias so that the idling current is "negative".
But again, this will never happen when playing music, there is plenty of time between transients to recover.
But it is an issue when trying to measure the performance/properties on the bench.
![gainA1](https://github.com/user-attachments/assets/c5e7ab45-62c0-45f5-b3b8-4e3bca6db670)
![outA1_3ms](https://github.com/user-attachments/assets/a34f3a11-9167-4e15-b82b-f05d67d6d20d)
![outA1_5s](https://github.com/user-attachments/assets/1b3454a4-0eb7-4ca7-882f-719a9878993d)
[Arkiv.zip](https://github.com/user-attachments/files/17398632/Arkiv.zip)

**Comment by tomhajjar** [comment]
"But again, this will never happen when playing music, there is plenty of time between transients to recover."

Famous last words...

Run the amp at high levels at a party and the Heatsink may be unable to keep the output devices within their SOA. Even amps with "proper" thermal compensation will pop if the heatsink is unable to shed heat and the amp goes into thermal runaway.

**Comment by bnilsson11** [comment]
On the other hand, the way this amp is designed, running at high levels will reduce the idling current bias to nothing.
It will sound like crap and will be turned off, if someone is sober enough to hear it.
I think it is at its best playing string quartets.

**Comment by tomhajjar** [comment]
Do you build amps? There are many designs over on diyAudio that have been simulated in LTspice. I have converted a few to Qucs-S.

I posted a few in the Discussion section.

**Comment by bnilsson11** [comment]
Yes, I do.
I also repair some, and then it is very useful to have a simulated version as a help in troubleshooting.
I will take a look, thanks.

**Comment by tomhajjar** [comment]
There are ways to convert LTspice schematics to Qucs but it is very time consuming...

**Comment by bnilsson11** [comment]
A comment about temperature: If I start to add some bells and whistles to the A1 design, a lot of those possible extensions will have to be checked for temperature stability. E.g. a constant current diode is definitely not without a temperature coefficient. Real estate is not unlimited, especially if I want to restrict myself to the credit card form factor of the ebay variant, and temperature compensation will take some extra space.

**Comment by tomhajjar** [comment]
Your simulations will be flawed since virtually every device is generating heat and will be hotter than ambient. 

LTspice supports most manufacturer's Spice "thermal" models and ngspice supports some manufacturer's "thermal" models. These Spice models take into account self-heating. Not sure if 2Nxx55 thermal models exist. Below is some of the work I have done. The guys over on the ngspice discussions would know what is supported.

https://github.com/ra3xdh/qucs_s/discussions/977

https://forum.qorvo.com/t/models-of-transistors-with-thermal-parameters/15809

**Comment by bnilsson11** [comment]
Simulations are not reality, can only give a hint of where possible problems can be expected.

A follow-up question to Parameter Sweep TEMP:

The Param Sweep TEMP 0-100 crashes for me, consistently at around 53 degrees with the usual

> doAnalyses: DC:  Timestep too small;  temp = 52.92: trouble with xlm7912.x1:j_2n3458_n-instance j.xlm7912.x1.jt1

On the other hand, a manual "Calculate DC bias" has absolutely no problem at any temperature in this range.
Why is this?
Can it be corrected?
Of course I can do "Calculate DC bias " 100 times to make a graph, but I would prefer not.

**Comment by tomhajjar** [comment]
Post your files.

For Manufacturers "real" thermal models, they will "blow-up" if the die temperatures exceed set limits.

**Comment by bnilsson11** [comment]
An attempt to keep the fundamental audible properties of the MFA1, but eliminating the current control feedback drawbacks.
A voltage-limited window of±0.6V and a subsequent integrator will allow the current sense voltage of 0.6A*0.33Ohm to be used, and the load current drift will be shaved off.
In this version there is a serious design mistake, see if you can spot it. It is all about the thermal behaviour.
Enjoy...

[MFA1v9dbg_prj.zip](https://github.com/user-attachments/files/17480709/MFA1v9dbg_prj.zip)

**Comment by tomhajjar** [comment]
Schematic has become way too "busy". Adding parts like voltage regulators with complicated transient models doesn't help.
Also why add the OpAmps Servo's instead of caps?

**Comment by bnilsson11** [comment]
Yes, it is busy.
It is the way to catch unexpected interaction between parts. 
If I want to simulate the startup behaviour for excessive currents that could blow the fuse, it is useful to include the 7812/7912 and turn off the initial DC op in TRANS, I got some interesting revelations there. 
If I really want a detailed TRANS I to study the signal details replace the 7912/7812 by 12V sources to reduce the overhead and include the TRANS DC op again, of course.

"Calculate DC bias" is not a TRANS nor AC, it is just strange that it cannot be automated with TEMP here since the manual DC bias works without problems for any TEMP within 0-100. And that was the actual question.


The design flaw I mentioned is that you cannot use the regulators as "voltage references" cancelling out to zero after adjustment, since the temperature coefficient is not the same for 78xx and 79xx regulators. The carefully trimmed zero DC offset will drift away quickly once the temperature is changed. It took this kind of "complete" simulation to realise that.
Changing to a dual diode and using the forward 0.6V as reference voltage works much better. It is 1) predictable, 2) same temperature for both sides since it is on the same chip, 3) higher ambient temperature will reduce the set point for the idling current and thus will make it safer. 

I use the integrator for practical reasons. I need an opamp to create the sense window function, but then I get the wrong sign for the feedback. A second opamp is needed to restore it, and why not use it as an integrator? The time constant needed is actually 2-10 seconds, and this is difficult to do with a single cap.

**Comment by tomhajjar** [comment]
Expecting complicated Behavioral or Transient Spice models to function during power ramp up/dn is ill advised...

**Comment by bnilsson11** [comment]
Ok, understood.
Any solution to the TEMP sweep?

**Comment by tomhajjar** [comment]
Post a cleaned up schematic and maybe some of the guys can help.

**Comment by bnilsson11** [comment]
I now have a small clue about the TEMP≠27 issue, it seems like a strange convergence problem.
I removed the feedback loop and tweaked the set point by adding precisely adjusted voltage sources at appropriate places.
Then there was no anomaly around TEMP=27, and the circuit behaved normally, except for the removed feedback.

I restored the normal feedback condition and made a PARAM sweep TEMP 26-28, 2001 steps, and monitored the output DC offset as an indicator. See the attached graph.
![DCoffsetVsTEMP](https://github.com/user-attachments/assets/8ee6f21e-f356-4422-9066-fcb0741b35eb)
There is a correctly converging OP at exactly 27 degrees, where the output DC offset is near 0V as it should.
Outside of 27 degrees the OP arrives at some very strange latch-up condition, far from hardware reality.
This is baffling to say the least.

Maybe ngspice is using one algorithm at TNOM successfully, and a different algorithm at TEMP≠TNOM, that fails?

The reason I have never encountered this before is of course that I have not investigated temperature effects until now.

Do you have any advice on how to set up / tune the OP convergence?

I made a qucs-s installation on Windows11 to make sure the behaviour is not platform dependent. I got exactly the same result there.

**Comment by ivandi69** [comment]
I did a slow start simulation and it looks like an instability caused by the feedback. The difference between VC7 and VC8 at 26.85 degrees is odd. At 0s VC8 is at -165mV even if all sources are at 0V. Setting initial condition for VC8 or changing the solver to KLU or decreasing tstep doesn't solve the problem. Asking ngspice developers for help is the best option.
![image](https://github.com/user-attachments/assets/629628c7-a804-4830-9b1e-1d58b1eb29e4)
[MFA1.sch.txt](https://github.com/user-attachments/files/17557350/MFA1.sch.txt)

**Comment by tomhajjar** [comment]
Numerous threads on diyAudio about this amp. None are favorable...

If you want help from the ngspice guys you need to clean up the amp and make a netlist including plot statements. You can use DuSpiceStart.exe to confirm netlist works. Present your issues and maybe someone can help. 

If all else fails make a version in LTspice and ask them for assistance. There may be a schematic on diyAudio already done.
https://hobby-electronics.ro/viewtopic.php?t=254

**Reply by bnilsson11** [reply]
> Numerous threads on diyAudio about this amp. None are favorable...

Not surprising.
It plays music very well (if not pushed too far), but on the test bench it is a disaster.

**Comment by ivandi69** [comment, marked as answer]
Found it :)
R19 has TC1 set to 45. After setting it to 0 everything works fine:
![image](https://github.com/user-attachments/assets/438c61ad-a4c5-41b9-8c45-d2380bb1ec03)

**Comment by tomhajjar** [comment]
post your updated files

**Comment by ivandi69** [comment]
[MFA1-debug2.sch.txt](https://github.com/user-attachments/files/17560728/MFA1-debug2.sch.txt)

**Comment by tomhajjar** [comment]
Your schematic is different than the two I found. Look at the collectors of the two diff pairs...

![MFA1_pwramp](https://github.com/user-attachments/assets/c0f6ce17-9a65-444d-9c01-ce732281357c)
![mf-a1](https://github.com/user-attachments/assets/e9119b6a-ccd9-40ef-b48d-1da67761c54b)

**Comment by ivandi69** [comment]
It is not **my** schematic. It is the [schematic ](https://github.com/user-attachments/files/17369290/MFA1-debug.sch.zip) that @bnilsson11 has problems with. Anyway, it is solved now.

**Comment by bnilsson11** [comment]
Fantastic detective work, kudos for that.

It will remain a mystery how that could happen, though.
I have definitely not put that TC in, at least not knowingly.
But it is all in the past now…

Once again, thanks.

> 29 okt. 2024 kl. 17:36 skrev ivandi69 ***@***.***>:
> 
> 
> Found it :)
> R19 has TC1 set to 45. After setting it to 0 everything works fine:
> image.png (view on web) <https://github.com/user-attachments/assets/438c61ad-a4c5-41b9-8c45-d2380bb1ec03>
> —
> Reply to this email directly, view it on GitHub <https://github.com/ra3xdh/qucs_s/discussions/994#discussioncomment-11089948>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAZXLDWIO6SX5YONI4ZDVQTZ562QVAVCNFSM6AAAAABP5XO5QSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMBYHE4TIOA>.
> You are receiving this because you authored the thread.
> 

Bengt Nilsson
Ansåsvägen 8
43853 Hindås
Sweden
***@***.***

**Comment by tomhajjar** [comment]
@bnilsson11 

There is 2 errors in the schematic. Input diff pair collector...

**Reply by bnilsson11** [reply]
Yes, possibly.
However, I tested both ways some time back, and I found that I got slightly better PSRR if both collectors used the same voltage supply. But at that time I did not know how to simulate distorsion, so I should take a look at it again.
Thanks for the suggestion.

**Comment by tomhajjar** [comment]
Cleaned up the old schematic and added FFT.

![2024-10-30_111948](https://github.com/user-attachments/assets/4b85d2eb-1ca1-45ec-a88a-edc22b32edc5)
![2024-10-30_112146](https://github.com/user-attachments/assets/cb0aba3c-37df-4c8f-a6f2-b1871c6dc926)

[Amp_MFA1_prj.zip](https://github.com/user-attachments/files/17575199/Amp_MFA1_prj.zip)

**Comment by bnilsson11** [comment]
Great job, very interesting.However, I am travelling right now, back on monday. I have to refrain from a proper discussion until then.Thanks for all the help, to get the TEMP problem solved.MvhBengt NilssonSkickat från min iPhone30 okt. 2024 kl. 16:22 skrev tomhajjar ***@***.***>:﻿
Cleaned up the old schematic and added FFT.
2024-10-30_111948.jpg (view on web)
2024-10-30_112146.jpg (view on web)
Amp_MFA1_prj.zip

**Reply by bnilsson11** [reply]
I tried it and found it very useful, new insights.
Thanks!

**Reply by bnilsson11** [reply]
I have a question about "OpAmp_ac_trans_fft".
I wanted to see what the FFT graph of a "clean, undistorted" source would look like.
If I change the nutmeg script "NutmegEq2" from "S_dB = db(v(out))" to "S_dB = db(v(in))", the FFT graph looks the same, no change. This is surprising.
But if I disable the input resistor R2 to disconnect the source from the circuit, the FFT is different. Now it looks more ideal.
Can this be explained?

![fftOut](https://github.com/user-attachments/assets/7a2cbb45-aee1-4784-8727-068425d885a4)
![fftIn](https://github.com/user-attachments/assets/0ac609f3-8f5b-4cce-b101-2bf19cff2f62)
![fftInNC](https://github.com/user-attachments/assets/c3f33c70-ac73-4511-b06a-fd03362a003c)

**Comment by tomhajjar** [comment]
You can experiment with the FFT parameters to get a plot closer to what you expect. 

![2024-11-07_130952](https://github.com/user-attachments/assets/074fe64e-2072-4d03-a3e9-72ca44b52950)

[OP491_trans_fft.zip](https://github.com/user-attachments/files/17667652/OP491_trans_fft.zip)

**Reply by bnilsson11** [reply]
I think you missed my point.
I found it surprising that
1) the input spectrum was the same as the output spectrum for this circuit;
2) the input spectrum was as expected (as the third of my previous graphs) only if it was disconnected from the circuit.
With "Expected" I mean a single 1kHz peak without harmonics.;

However, if I use 10V source amplitude to induce clipping, in and out spectrum are indeed different, so there is no mixup in what is plotted.

It seems like something is leaking back into the source from the circuit, and the "ideal" voltage source is not "ideal" at all.
With "ideal" I mean zero output impedance, any signal back flow should be grounded by the source.

Try it and you will understand what I mean.
It is more obvious if you use 1Hz BW in the FFT.
The first posted image here is for the IN source only, no circuit connected. Looks like an "ideal" 1kHz source as defined above.
The second is IN with the input resistor connected. It has unexpected harmonics.
Again, try for yourself in your schematic. It is easier that way, less risk of misunderstandings.
Maybe this is normal, maybe there are no "ideal"sources even in simulations? What is the output impedance of this "ideal" source?
![fftInNC1Hz](https://github.com/user-attachments/assets/746b4c04-f5b2-4e37-b504-e4b542de5f92)
![fftIn1Hz](https://github.com/user-attachments/assets/2b80b466-7ec6-4d91-a827-b86cb93a5903)

**Comment by ivandi69** [comment]
This is a quick test. No 1kHz in o2 output and no 2kHz in o1 output.
![image](https://github.com/user-attachments/assets/b6b8e38a-024c-4615-a643-1d3674b90199)

**Reply by bnilsson11** [reply]
I re-created your quick-test setup, and I got similar results as you but not the low level noise structure in the graphs.
Is this platform dependent? I use ngspice 42.
![ivan](https://github.com/user-attachments/assets/6e77fb1a-1cf6-46df-8309-9a086d5e38fa)

**Comment by felix-salfelder** [comment]
Here's a wild guess. The (noise in the) spectrum depends on the time stepping. Disconnecting the source may have changed the spectrum at the source just by shuffling the sampling points.

To see the effect, put in two identical sources, connect only one to the circuit and look at the spectrum at the output of the unconnected one.

Another important noise source, warmup. Not sure how it applies here, but generally the first few cycles should go to the bin before doing anything with fft.

**Reply by bnilsson11** [reply]
I added one more source (2kHz named in0) to the circuit, grounded it but I did not connect it to the input.
And indeed, there was noise on the unconnected source as well.
I also replaced the opamp to TL081, but there was no change in the behaviour, so I changed back to OP491.
What conclusions can be made from this?
![OpAmp_ac_trans_fft](https://github.com/user-attachments/assets/f908aaf1-d0a0-4d03-ba0c-fdaecc4bf563)

**Comment by felix-salfelder** [comment]
The amplitude of the noise seems pretty high. This could be a bug, or a flaw far down under the hood in one of the algorithms. Maybe there is a hack that masks the problem. But beware, it could also be a side effect of the hack for a previous one.

"5.3.1.4 Error Due to Interpolation" and "5.3.1.5 Error Due to Simulator Noise" in the "Designers Guide to Spice and _another_)" gives some generic background (TW: plug for _another_).

> What conclusions can be made from this?

The high level observation is that tools that shine on simple examples do not necessarily solve all problems well. My personal conclusion is that we need to move on, reduce complexity, standardise, modularise, exchange data between tools, identify those that do a solid job.

**Comment by tomhajjar** [comment]
Questions about FFT can be asked over at ngspice...

https://sourceforge.net/p/ngspice/discussion/127605/thread/03b82f8bf2/

**Reply by bnilsson11** [reply]
Please check Please check https://sourceforge.net/p/ngspice/discussion/133842/thread/3e6023d8eb/
The answer was a bit inconclusive to me, maybe you can figure it out.

**Comment by bnilsson11** [comment]
Agreed, this is something for ngspice, not neccessarily qucs-s.MvhBengt NilssonSkickat från min iPhone8 nov. 2024 kl. 13:53 skrev tomhajjar ***@***.***>:﻿
Questions about FFT can be asked over at ngspice...
https://sourceforge.net/p/ngspice/discussion/127605/thread/03b82f8bf2/

**Reply by felix-salfelder** [reply]
Something went wrong with the netlist export and upload in the message board over at ngspice.

.. a reminder that the underlying [issue](https://github.com/Qucs/qucs/issues/1094) that renders Qucs(-S) files useless needs fixing.

**Reply by ra3xdh** [reply]
I have answered here: https://github.com/ra3xdh/qucs_s/issues/1047#issuecomment-2466796479 Please don't post messages about file formats in this thread. It is not related to the original subject. You may create a new discussion about file formats if you wish. And don't throw words about "useless" files.

**Comment by ivandi69** [comment]
Here is an example WRT [Unexpected FFT result](https://sourceforge.net/p/ngspice/discussion/133842/thread/3e6023d8eb/):
![image](https://github.com/user-attachments/assets/875522b2-e9cf-4793-a080-34f94cf66ff2)
[fft.sch.txt](https://github.com/user-attachments/files/17691427/fft.sch.txt)
